src/core_plugins/tagcloud/public/tag_cloud_visualization.js

```diff
@@ -46,8 +46,9 @@ export class TagCloudVisualization {
       if (!this._bucketAgg) {
         return;
       }
-      const filter = this._bucketAgg.createFilter(event);
-      this._vis.API.queryFilter.addFilters(filter);
+      this._vis.API.events.addFilter(
+        this._data, 0, this._data.rows.findIndex(row => row[0] === event)
```

Comment:
I would suggest instead of trying to find the row, we change the tagcloud visualization to attach that data and pass it when clicking a tag:

```diff
diff --git a/src/core_plugins/tagcloud/public/tag_cloud.js b/src/core_plugins/tagcloud/public/tag_cloud.js
index db3b1a8506..3d158ec4d5 100644
--- a/src/core_plugins/tagcloud/public/tag_cloud.js
+++ b/src/core_plugins/tagcloud/public/tag_cloud.js
@@ -224,7 +224,7 @@ class TagCloud extends EventEmitter {
       const self = this;
       enteringTags.on({
         click: function (event) {
-          self.emit('select', event.rawText);
+          self.emit('select', event);
         },
         mouseover: function () {
           d3.select(this).style('cursor', 'pointer');
diff --git a/src/core_plugins/tagcloud/public/tag_cloud_visualization.js b/src/core_plugins/tagcloud/public/tag_cloud_visualization.js
index e9cc02ee9e..0fb5cc11ee 100644
--- a/src/core_plugins/tagcloud/public/tag_cloud_visualization.js
+++ b/src/core_plugins/tagcloud/public/tag_cloud_visualization.js
@@ -47,7 +47,7 @@ export class TagCloudVisualization {
         return;
       }
       this._vis.API.events.addFilter(
-        this._data, 0, this._data.rows.findIndex(row => row[0] === event)
+        this._data, 0, event.meta.rowIndex
       );
     });
     this._renderComplete$ = Observable.fromEvent(this._tagCloud, 'renderComplete');
@@ -121,12 +121,15 @@ export class TagCloudVisualization {
       this._bucketAgg = null;
     }
 
-    const tags = data.rows.map(row => {
+    const tags = data.rows.map((row, rowIndex) => {
       const [tag, count] = row;
       return {
         displayText: this._bucketAgg ? this._bucketAgg.fieldFormatter()(tag) : tag,
         rawText: tag,
-        value: count
+        value: count,
+        meta: {
+          rowIndex,
+        },
       };
     });
```

Reply:
Maybe we should even attach the data to each tag's `meta` and use that one instead of writing it to `this._data`, which I think could potentially in very bad timing not be the `data` that was used to render the tag anymore, since tag cloud does render async.

Reply:
i.e.

```diff
diff --git a/src/core_plugins/tagcloud/public/tag_cloud.js b/src/core_plugins/tagcloud/public/tag_cloud.js
index db3b1a8506..3d158ec4d5 100644
--- a/src/core_plugins/tagcloud/public/tag_cloud.js
+++ b/src/core_plugins/tagcloud/public/tag_cloud.js
@@ -224,7 +224,7 @@ class TagCloud extends EventEmitter {
       const self = this;
       enteringTags.on({
         click: function (event) {
-          self.emit('select', event.rawText);
+          self.emit('select', event);
         },
         mouseover: function () {
           d3.select(this).style('cursor', 'pointer');
diff --git a/src/core_plugins/tagcloud/public/tag_cloud_visualization.js b/src/core_plugins/tagcloud/public/tag_cloud_visualization.js
index e9cc02ee9e..e3e4bee1dd 100644
--- a/src/core_plugins/tagcloud/public/tag_cloud_visualization.js
+++ b/src/core_plugins/tagcloud/public/tag_cloud_visualization.js
@@ -47,7 +47,7 @@ export class TagCloudVisualization {
         return;
       }
       this._vis.API.events.addFilter(
-        this._data, 0, this._data.rows.findIndex(row => row[0] === event)
+        event.meta.data, 0, event.meta.rowIndex
       );
     });
     this._renderComplete$ = Observable.fromEvent(this._tagCloud, 'renderComplete');
@@ -121,12 +121,16 @@ export class TagCloudVisualization {
       this._bucketAgg = null;
     }
 
-    const tags = data.rows.map(row => {
+    const tags = data.rows.map((row, rowIndex) => {
       const [tag, count] = row;
       return {
         displayText: this._bucketAgg ? this._bucketAgg.fieldFormatter()(tag) : tag,
         rawText: tag,
-        value: count
+        value: count,
+        meta: {
+          data,
+          rowIndex,
+        },
       };
     });
 ```